app/org/sagebionetworks/bridge/dynamodb/DynamoExternalIdDao.java

```diff
@@ -84,163 +84,140 @@ public ExternalIdentifier getExternalId(StudyIdentifier studyId, String external
-    public void addExternalIds(StudyIdentifier studyId, List<String> externalIds) {
-        checkNotNull(studyId);
-        checkNotNull(externalIds);
+    public void createExternalId(ExternalIdentifier externalId) {
+        checkNotNull(externalId);
+        
+        mapper.save(externalId);
+    }
 
-        // We validate a wider range of issues in the service, but check size again because this is 
-        // specifically a database capacity issue.
-        if (externalIds.size() > addLimit) {
-            throw new BadRequestException("List of externalIds is too large; size=" + externalIds.size() + ", limit=" + addLimit);
-        }
-        List<DynamoExternalIdentifier> idsToSave = externalIds.stream().map(id -> {
-            return new DynamoExternalIdentifier(studyId, id);
-        }).filter(externalId -> {
-            return mapper.load(externalId) == null;
-        }).collect(Collectors.toList());
+    @Override
+    public void deleteExternalId(ExternalIdentifier externalId) {
+        checkNotNull(externalId);
         
-        if (!idsToSave.isEmpty()) {
-            List<FailedBatch> failures = mapper.batchSave(idsToSave);
-            BridgeUtils.ifFailuresThrowException(failures);
-        }
+        mapper.delete(externalId);
     }
     
+    /**
+     * There are two methods in the ExternalIdService that are intended to be called in order: 
+     * beginAssignExternalId() which will return an external ID, and commitAssignExternalId() 
+     * which calls the DAO to persist that external ID in the context of an account transaction. 
+     * This allows us to abandon a database update if the DynamoDB update fails. The whole thing 
+     * is then wrapped in a catch statement that undoes the DynamoDB update if the database 
+     * update fails. This method also checks a couple of things before these persistence calls: 
+     * that a health code is assigned and that the external ID does exist.
+     */
     @Override
-    public void assignExternalId(StudyIdentifier studyId, String externalId, String healthCode) {
-        checkNotNull(studyId);
-        checkArgument(isNotBlank(externalId));
-        checkArgument(isNotBlank(healthCode));
-        
-        DynamoExternalIdentifier keyObject = new DynamoExternalIdentifier(studyId, externalId);
-        DynamoExternalIdentifier identifier = mapper.load(keyObject);
-
-        // If the identifier doesn't exist, or the same code has already been set, do nothing
-        if (identifier != null && !healthCode.equals(identifier.getHealthCode())) {
+    public void commitAssignExternalId(ExternalIdentifier externalId) {
+        if (externalId != null) {
+            DynamoExternalIdentifier key = new DynamoExternalIdentifier(
+                    externalId.getStudyId(), externalId.getIdentifier());
+            if (externalId.getHealthCode() == null || mapper.load(key) == null) {
+                throw new IllegalStateException("Invalid external ID assignment.");
```

Comment:
My thinking is that it would be a programmer error to call the commit method without having verified that the external ID exists and is unassigned. But the health code could have been assigned concurrently, so changed to a ConcurrentModificationException (which is a 409).